xenmgr/XenMgr/PowerManagement.hs

```diff
@@ -255,6 +261,19 @@ shutdownCommon offCommand force = do
       -- wait for actuall poweroff
       threadDelay $ 360 * (10^6)
 
+-- Got idle shutdown trigger from xcpmd. Start a thread
+-- that will force the host off after 5 minutes. While that's running,
+-- perform the normal shutdown procedure passed in via offCommand
+shutdownIdle :: String -> XM ()
+shutdownIdle offCommand = do
+    liftIO $ forkIO $ (threadDelay (300 * (10^6)) >> forceShutdown)
+    shutdownCommon offCommand False
+  where
+    forceShutdown = do
+      info "PM: 5 minute cap reached for idle shutdown, performing immediate host shutdown"
+      spawnShell "poweroff -ff"
```

Comment:
I think a VM not shutting down would be the biggest cause of a shutdown hang.  I was expecting to see the timeout trigger destroy for all VMs and then a normal shutdown.  `poweroff -ff` bypasses the shutdown scripts and that seems too aggressive to me - like we aren't guaranteed that rsyslog & the kernel will have flushed the output to disk.   But I can also see `poweroff -ff` being potentially necessary to ensure we really do shutdown for some other process/daemon hanging.  Still, xenmgr is stopped fairly early in the shutdown sequence, so if we really want that guarantee, it seems like xenmgr should fork & detach a shell script to run `sleep X && poweroff -ff` while letting the rest of the shutdown continue.  Am I overthinking this?

Reply:
This implementation came out of an overwhelming desire to make sure the system is completely powered down within 5 minutes of the initial request.  It was acceptable to bypass proper/graceful shutdown sequences to meet this goal.  We've seen guest VMs not fully cleaned up with ```xl destroy``` and they leave behind stuck ```tapdisk``` processes which block the system from shutting off normally.

>  But I can also see `poweroff -ff` being potentially necessary to ensure we really do shutdown for some other process/daemon hanging.  Still, xenmgr is stopped fairly early in the shutdown sequence, so if we really want that guarantee, it seems like xenmgr should fork & detach a shell script to run `sleep X && poweroff -ff` while letting the rest of the shutdown continue.

That might be the better implementation, you're not overthinking it.  ```forkIO``` is still a Thread so it's ultimately bound by the main XenMgr process.  Depending on which dom0 PID is stuck we may be able to get further along in the shutdown before needing to rely on the `poweroff -ff` bomb we left behind.

Reply:
Ok, thanks for the info.  Since this meets your requirements we can go with it and improve later if necessary.